pkg/controller/keyvault/gateway_kv_serviceaccount.go

```diff
@@ -0,0 +1,137 @@
+package keyvault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/controllername"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/metrics"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+var kvSaControllerName = controllername.New("gateway", "keyvault", "serviceaccount")
+
+// KvServiceAccountReconciler reconciles the default "azure-app-routing-kv" ServiceAccount in each namespace where users
+// create a Gateway resource with the Keyvault cert TLS option. Users can tie this ServiceAccount to their own MSI via
+// workload identity through annotations on their Gateway resources.
+type KvServiceAccountReconciler struct {
+	client client.Client
+	events record.EventRecorder
+}
+
+func NewKvServiceAccountReconciler(mgr ctrl.Manager) error {
+	metrics.InitControllerMetrics(kvSaControllerName)
+
+	return kvSaControllerName.AddToController(ctrl.NewControllerManagedBy(mgr).
+		For(&gatewayv1.Gateway{}), mgr.GetLogger()).
+		Complete(&KvServiceAccountReconciler{
+			client: mgr.GetClient(),
+			events: mgr.GetEventRecorderFor("app-routing-operator"),
+		})
+
+}
+
+func (k *KvServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res := ctrl.Result{}
```

Comment:
I know this is the pattern some other controllers use but I don't really like it. Named returns are better and accomplish the desired defer behavior better

Reply:
to clarify named return is better because you can follow go conventions instead of always having to remember to use the res and err vars. It's easy to mess up the current pattern because it relies on always using these exact vars

Reply:
yeah good call I'll move over to named returns